__device-tests__/gutenberg-editor-sanity-test-4-visual.test.js

```diff
@@ -0,0 +1,235 @@
+			await editorPage.setHtmlContent( buttonCustomColors );
+
+			const buttonsBlock = await editorPage.getBlockAtPosition(
+				blockNames.buttons
+			);
+			await buttonsBlock.click();
+
+			const buttonBlock = await editorPage.getBlockAtPosition(
+				blockNames.button
+			);
+			await buttonBlock.click();
+
+			// Get button's block TextInput
+			const buttonBlockTextInput = await editorPage.getButtonBlockTextInputAtPosition();
+			await selectTextFromElement(
+				editorPage.driver,
+				buttonBlockTextInput
+			);
+
+			// Visual test check
+			const screenshot = await takeScreenshot( {
+				withoutKeyboard: true,
+			} );
+			expect( screenshot ).toMatchImageSnapshot();
+
+			await buttonBlockTextInput.click();
+
+			await editorPage.removeBlockAtPosition( blockNames.button );
+		} );
+
+		it( 'Edit text styles', async () => {
+			await editorPage.addNewBlock( blockNames.buttons );
+
+			const buttonsBlock = await editorPage.getBlockAtPosition(
+				blockNames.buttons
+			);
+
+			const firstButtonTextInput = await editorPage.getButtonBlockTextInputAtPosition();
+			await editorPage.typeTextToTextBlock(
+				firstButtonTextInput,
+				e2eTestData.listItem2
+			);
+			await selectTextFromElement(
+				editorPage.driver,
+				firstButtonTextInput
+			);
+
+			// Toggle Bold formatting for the first Button block
+			await editorPage.toggleFormatting( 'Bold' );
+
+			// Add a second button block
+			await editorPage.addButtonWithInlineAppender( 2 );
+			const secondButtonTextInput = await editorPage.getButtonBlockTextInputAtPosition(
+				2
+			);
+			await editorPage.typeTextToTextBlock(
+				secondButtonTextInput,
+				e2eTestData.listItem2
+			);
+			await selectTextFromElement(
+				editorPage.driver,
+				secondButtonTextInput
+			);
+
+			// Toggle Italic formatting for the first Button block
+			await editorPage.toggleFormatting( 'Italic' );
+
+			// Add a third button block
+			await editorPage.addButtonWithInlineAppender( 3 );
+			const thirdButtonTextInput = await editorPage.getButtonBlockTextInputAtPosition(
+				3
+			);
+			await editorPage.typeTextToTextBlock(
+				thirdButtonTextInput,
+				e2eTestData.listItem2
+			);
+			await selectTextFromElement(
+				editorPage.driver,
+				thirdButtonTextInput
+			);
+
+			// Toggle Strikethrough formatting for the first Button block
+			await editorPage.toggleFormatting( 'Strikethrough' );
+
+			// Unfocus the Buttons block
+			const titleElement = await editorPage.getTitleElement( {
+				autoscroll: true,
+			} );
+			await titleElement.click();
+
+			await editorPage.dismissKeyboard();
+
+			// Visual test check for unselected state
+			const screenshot = await takeScreenshot();
+			expect( screenshot ).toMatchImageSnapshot();
+
+			await buttonsBlock.click();
+			await editorPage.removeBlockAtPosition( blockNames.buttons );
+		} );
+
+		it( 'Link from the clipboard is presented as an option in the link picker', async () => {
```

Comment:
For this case, since we are testing copying a link from the clipboard and expecting it to show in the link picker it is very attached to the device imagine something fails like a new OS permission is added and the link can't be pasted, and having a visual test of the whole workflow would allow us not to have to test this manually.

Reply:
Fair point. Avoiding a clipboard API mock provides additional coverage (e.g. OS permissions). There a few approaches to testing the spirit of this test case we could consider: 

1. Capture visual images within an e2e tests. 
2. Assert element presence within an e2e tests. 
3. Mock the clipboard API and assert element presence within an integration tests. 

It seems you prefer the first two for this specific test case, which I think is reasonable. 

Additionally, I do not know if there is a meaningful "cost" between the first two. The first may "cost" more in regards to test speed or repository size for stored images. While the test case description doesn't necessarily _require to visual testing_ (we could just assert the element presence), it is additional coverage. I suppose it is a judgement call as to whether the visual coverage is worth it.